walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,92 @@
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  //public static final int START_OF_DAY = TimeRange.getTimeInMinutes(0, 0);
+  private static final int END_OF_DAY = TimeRange.getTimeInMinutes(23, 59);
+  //private static final int WHOLE_DAY = TimeRange.getTimeInMinutes(0, 24 * 60);
+  
+  private static final Comparator<Event> ORDER_BY_START = new Comparator<Event>() {
+    @Override
+    public int compare(Event a, Event b) {
+      return Long.compare(a.getWhen().start(), b.getWhen().start());
+    }};
+    
+  private List<TimeRange> acceptableMeetingTimes = new ArrayList<TimeRange>();
+  
+  public Collection<TimeRange> query(Collection<Event> eventCollection, MeetingRequest request) {
+    List<String> meetingAttendees = new ArrayList<String>(request.getAttendees());
+    List<Event> events = new ArrayList<Event>(eventCollection);
+    List<Event> importantEvents = new ArrayList<Event>();  
+    
+    // Too long of a request
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return acceptableMeetingTimes; 
+    }
+    
+    // No events or attendees
+    if (events.isEmpty() || meetingAttendees.isEmpty()) {
+      acceptableMeetingTimes.add(
+        TimeRange.WHOLE_DAY); 
+      return acceptableMeetingTimes;
+    }
+    
+    // Ignores People not Attending
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        importantEvents.add(event); 
+      }
+    }
+    
+    // If the list of important events is empty, return 
+    if (importantEvents.isEmpty()) {
+      acceptableMeetingTimes.add(TimeRange.WHOLE_DAY); 
+      return acceptableMeetingTimes; 
+    }
+    
+    // Order the events by their start time 
+    Collections.sort(importantEvents, ORDER_BY_START); 
+    
+    // Add the event that starts first
+    acceptableMeetingTimes.add(TimeRange.fromStartEnd(0, importantEvents.get(0).getWhen().start(), false)); 
+
+    // Set the end time and start time as the end of the first event 
+    int end = importantEvents.get(0).getWhen().end();
+     
+    // Iterate through the rest of the events 
+    for (Event event : importantEvents) {
+      int start = event.getWhen().start(); 
+      if (event.getWhen().start() >= end) {
+        end = event.getWhen().end();  
+      }
+      
+      if (checkMeetingDuration(TimeRange.fromStartEnd(end, start, false), request.getDuration())) {
+        acceptableMeetingTimes.add(TimeRange.fromStartEnd(end, start, false)); 
+        end = event.getWhen().end();
+      }
+    }
+    
+    // If there is still time between the last meeting and the end of the day, add the time range
+    if (checkMeetingDuration(TimeRange.fromStartEnd(end, END_OF_DAY, false), request.getDuration())) {
+      acceptableMeetingTimes.add(TimeRange.fromStartEnd(end, END_OF_DAY, true));
```

Comment:
I think this needs to be a loop of some sort, because you could fit some number of meetings between end-of-last-meeting and end-of-day, probably.

Reply:
Done